nemo/src/io/formats/sparql.rs

```diff
@@ -0,0 +1,186 @@
+                .to_u64()
+                .and(Some(()))
+                .ok_or(ValidationErrorKind::ImportExportLimitNegative),
+            SparqlParameter::Base => Ok(()),
+            SparqlParameter::Format => DsvValueFormats::try_from(value).and(Ok(())).map_err(|_| {
+                ValidationErrorKind::ImportExportValueFormat {
+                    file_format: "rdf".to_string(),
+                }
+            }),
+            SparqlParameter::IgnoreHeaders => Ok(()),
+            SparqlParameter::Endpoint => value
+                .to_iri()
+                .ok_or(ValidationErrorKind::ImportExportInvalidIri)
+                .and_then(|iri| {
+                    Iri::parse(iri)
+                        .map(|_| ())
+                        .map_err(|_| ValidationErrorKind::ImportExportInvalidIri)
+                }),
+            SparqlParameter::Query => value
+                .to_plain_string()
+                .ok_or(ValidationErrorKind::ImportExportLimitNegative)
+                .and_then(|query| {
+                    Query::parse(query.as_str(), None).map(|_| ()).map_err(|e| {
+                        ValidationErrorKind::ImportExportInvalidSparqlQuery {
+                            oxi_error: e.to_string(),
+                        }
+                    })
+                }),
+        }
+    }
+}
+
+#[derive(Clone)]
+pub(crate) struct SparqlBuilder {
+    limit: Option<u64>,
+    delimiter: u8,
+    value_formats: Option<DsvValueFormats>,
+    ignore_headers: bool,
+    parsed_query: Query,
+}
+
+impl From<SparqlBuilder> for AnyImportExportBuilder {
+    fn from(value: SparqlBuilder) -> Self {
+        AnyImportExportBuilder::Sparql(value)
+    }
+}
+impl FormatBuilder for SparqlBuilder {
+    type Parameter = SparqlParameter;
+    type Tag = SparqlTag;
+    fn new(
+        _tag: Self::Tag,
+        parameters: &Parameters<SparqlBuilder>,
+        _direction: Direction,
+    ) -> Result<(Self, Option<ResourceSpec>), ValidationErrorKind> {
+        // Copied from DsvBuilder
+        let value_formats = parameters
+            .get_optional(SparqlParameter::Format)
+            .map(|value| DsvValueFormats::try_from(value).unwrap());
+
+        let limit = parameters
+            .get_optional(SparqlParameter::Limit)
+            .map(|value| value.to_u64_unchecked());
+
+        let ignore_headers = parameters
+            .get_optional(SparqlParameter::IgnoreHeaders)
+            .as_ref()
+            .map(AnyDataValue::to_boolean_unchecked)
+            .unwrap_or(false);
+
+        // Sparql-specific fields
+        let endpoint = Iri::parse_unchecked(
+            parameters
+                .get_required(SparqlParameter::Endpoint.into())
+                .to_iri_unchecked(),
+        );
+
+        let query = parameters
+            .get_required(SparqlParameter::Query.into())
+            .to_plain_string_unchecked();
+
+        let parsed_query = Query::parse(query.as_str(), None).unwrap();
```

Comment:
This needs proper error handling, otherwise we'll just panic on syntax errors in the query. Generally, `unwrap()` is only ever acceptable in tests, at the very least, there should be an `expect()`.

Reply:
Well, if it had been `Query::parse(query.as_str(), None).expect("query has already been validated")`, that would have been clear to me without having to keep the whole context in my head.

Reply:
If the query has already been parsed before, then why do we have to do this again instead of keeping the result from the first time we did it?